iowrappers/users.go

```diff
@@ -79,25 +79,90 @@ func (r *RedisClient) UpdateSearchHistory(ctx context.Context, location string,
 		}
 	}
 
-	if reflect2.IsNil(userView.Favorites.SearchHistory) {
-		userView.Favorites = &user.PersonalFavorites{SearchHistory: make(map[string]user.LastSearchRecord)}
-	}
+	redisKey := strings.Join([]string{UserSearchHistoryPrefix, "user", userView.ID}, ":")
+	err := r.Get().Watch(ctx, func(tx *redis.Tx) error {
+		record := &user.LastSearchRecord{}
+		val, err := tx.HGet(ctx, redisKey, location).Result()
+		if err != nil && !errors.Is(err, redis.Nil) {
+			return err
+		}
+
+		if val != "" {
+			if err = json.Unmarshal([]byte(val), record); err != nil {
+				return fmt.Errorf("failed to unmarshal user search history: %w", err)
+			}
+		}
 
-	if _, exists := userView.Favorites.SearchHistory[location]; !exists {
-		userView.Favorites.SearchHistory[location] = user.LastSearchRecord{
-			Location:            location,
-			Count:               0,
-			LastSearchTimestamp: time.Now().Format(time.RFC3339),
+		// Update the record
+		record.Location = location
+		record.LastSearchTimestamp = time.Now().Format(time.RFC3339)
+		record.Count++
+
+		// Serialize the record
+		serialized, err := json.Marshal(record)
+		if err != nil {
+			return fmt.Errorf("failed to serialize user search history: %w", err)
 		}
+
+		// Execute the transaction
+		_, err = tx.TxPipelined(ctx, func(pipe redis.Pipeliner) error {
+			pipe.HSet(ctx, redisKey, location, serialized)
+			return nil
+		})
+		return err
+	}, redisKey)
+
+	if err != nil {
+		return fmt.Errorf("failed to update search history: %w", err)
 	}
 
-	data := userView.Favorites.SearchHistory[location]
-	data.Count++
-	data.LastSearchTimestamp = time.Now().Format(time.RFC3339)
-	userView.Favorites.SearchHistory[location] = data
+	return nil
+}
 
-	Logger.Debugf("updating user favorite for user %s: %+v", userView.ID, data)
-	return r.UpdateUser(ctx, userView)
+type Favorites struct {
+	MostFrequentSearch string `json:"most_frequent_search"`
+}
+
```

Comment:
lg, what are the two new functions doing now ? `UserFavorites` and `findTopLocation` ?

Reply:
User favorite finds the most searched location of a user. From users perspective they will not see difference in the profile page.